velox/exec/HashBuild.cpp

```diff
@@ -178,12 +278,315 @@ void HashBuild::addInput(RowVectorPtr input) {
+      1, targetIncrement / (rows->fixedRowSize() + outOfLineBytesPerRow));
+  numSpillBytes_ = numSpillRows_ * outOfLineBytesPerRow;
+  return false;
+}
+
+void HashBuild::spillInput(RowVectorPtr& input) {
+  VELOX_CHECK_EQ(input->size(), activeRows_.size());
+
+  if (!spillEnabled() || !spiller_->isAnySpilled() ||
+      !activeRows_.hasSelections()) {
+    return;
+  }
+
+  const auto numInput = input->size();
+  prepareInputIndicesBuffers(numInput, spiller_->spilledPartitionSet());
+  computeSpillPartitions(input);
+
+  vector_size_t numSpillInputs = 0;
+  for (auto row = 0; row < numInput; ++row) {
+    const auto partition = spillPartitions_[row];
+    if (FOLLY_UNLIKELY(!activeRows_.isValid(row))) {
+      continue;
+    }
+    if (!spiller_->isSpilled(partition)) {
+      continue;
+    }
+    activeRows_.setValid(row, false);
+    ++numSpillInputs;
+    spillInputIndicesBuffers_[partition]
+        ->asMutable<vector_size_t>()[numSpillInputs_[partition]++] = row;
+  }
+  if (numSpillInputs == 0) {
+    return;
+  }
+
+  for (uint32_t partition = 0; partition < numSpillInputs_.size();
+       ++partition) {
+    const int numSpillInputs = numSpillInputs_[partition];
+    if (numSpillInputs == 0) {
+      continue;
+    }
+    VELOX_CHECK(spiller_->isSpilled(partition));
+    spillPartition(
+        partition, numSpillInputs, spillInputIndicesBuffers_[partition], input);
+  }
+  activeRows_.updateBounds();
+}
+
+void HashBuild::prepareInputIndicesBuffers(
+    vector_size_t numInput,
+    const SpillPartitionNumSet& spillPartitions) {
+  for (const auto& partition : spillPartitions) {
+    if (spillInputIndicesBuffers_[partition] == nullptr) {
+      spillInputIndicesBuffers_[partition] = allocateIndices(numInput, pool());
+    } else {
+      AlignedBuffer::reallocate<vector_size_t>(
+          &spillInputIndicesBuffers_[partition], numInput);
+    }
+  }
+  std::fill(numSpillInputs_.begin(), numSpillInputs_.end(), 0);
+}
+
+void HashBuild::computeSpillPartitions(const RowVectorPtr& input) {
+  if (hashes_.size() < activeRows_.size()) {
+    hashes_.resize(activeRows_.size());
+  }
+  const auto& hashers = table_->hashers();
+  for (auto i = 0; i < hashers.size(); ++i) {
+    auto& hasher = hashers[i];
+    if (hasher->channel() != kConstantChannel) {
+      hashers[i]->hash(activeRows_, i > 0, hashes_);
+    } else {
+      hashers[i]->hashPrecomputed(activeRows_, i > 0, hashes_);
+    }
+  }
+
+  spillPartitions_.resize(input->size());
+  for (auto i = 0; i < spillPartitions_.size(); ++i) {
+    spillPartitions_[i] = spiller_->hashBits().partition(hashes_[i]);
+  }
+}
+
+void HashBuild::spillPartition(
+    uint32_t partition,
+    vector_size_t size,
+    BufferPtr indices,
+    const RowVectorPtr& input) {
+  VELOX_DCHECK(spillEnabled());
+
+  if (isInputFromSpill()) {
+    spiller_->spill(partition, wrap(size, std::move(indices), input));
+  } else {
+    int spillChannel = 0;
+    for (const auto& channel : keyChannels_) {
```

Comment:
Do we need to run these loops for each partition?

Reply:
Good point. Thanks!